sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -104,6 +105,9 @@ trait CheckAnalysis {
               s"filter expression '${f.condition.prettyString}' " +
                 s"of type ${f.condition.dataType.simpleString} is not a boolean.")
 
+          case j @ Join(_, _, NaturalJoin(_), _) =>
+            failAnalysis(s"natural join not resolved.")
```

Comment:
if this happens, it must because join operator itself has some problems, but not caused by the `NaturalJoin` type. I think we should remove this case and let other checking rules to tell users what's wrong.